sklearn/metrics/classification.py

```diff
@@ -1994,3 +1994,123 @@ def brier_score_loss(y_true, y_prob, sample_weight=None, pos_label=None):
+    that can be structured as true or false, but is inappropriate for ordinal
+    variables which can take on three or more values.
+    Which label is considered to be the positive label is controlled via the
+    parameter pos_label, which defaults to 1.
+
+    Read more in the :ref:`User Guide <calibration>`.
+
+    Parameters
+    ----------
+    y_true : array, shape (n_samples,)
+        True targets.
+
+    y_prob : array, shape (n_samples,)
+        Probabilities of the positive class.
+
+    reducer: string, must be among 'l1', 'l2', 'max'
+        Aggregation method.
+
+    nbins: int, positive, optional (default=10)
+        Number of bins. Computation requires nbins full passes on both
+        y_prob and y_true but larger nbins can increase the accuracy of the
+        calibration loss, provided sufficient data in bins.
+
+    normalize : bool, optional (default=True)
+        If true, return the mean loss per sample.
+        Otherwise, return the sum of the per-sample losses.
+
+    sample_weight : array-like of shape = [n_samples], optional
+        Sample weights.
+
+    pos_label : int or str, default=None
+        Label of the positive class. If None, the maximum label is used as
+        positive class
+
+    Returns
+    -------
+    score : float
+        calibration loss
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from sklearn.metrics import calibration_loss
+    >>> y_true = np.array([0, 0, 0, 1] + [0, 1, 1, 1])
+    >>> y_pred = np.array([0.25, 0.25, 0.25, 0.25] + [0.75, 0.75, 0.75, 0.75])
+    >>> calibration_loss(y_true, y_pred, nbins=2, \
+                         reducer="sum")  # doctest: +ELLIPSIS
+    0.0
+    >>> calibration_loss(y_true, y_pred, nbins=2, \
+                         reducer="max")  # doctest: +ELLIPSIS
+    0.0
+    >>> y_true = np.array([0, 0, 0, 0] + [1, 1, 1, 1])
+    >>> calibration_loss(y_true, y_pred, nbins=2, \
+                         reducer="sum")  # doctest: +ELLIPSIS
+    0.25
+    >>> calibration_loss(y_true, y_pred, nbins=2, \
+                         reducer="max")  # doctest: +ELLIPSIS
+    0.25
+
+    References
+    ----------
+    [1] `Chuan Guo, Geoff Pleiss, Yu Sun, Kilian Q. Weinberger. On Calibration
+        of Modern Neural Networks. Proceedings of the 34th International
+        Conference on Machine Learning, PMLR 70:1321-1330, 2017.
+        <http://proceedings.mlr.press/v70/guo17a.html>`
+    """
+    y_true = column_or_1d(y_true)
+    y_prob = column_or_1d(y_prob)
+    assert_all_finite(y_true)
+    assert_all_finite(y_prob)
+    check_consistent_length(y_true, y_prob, sample_weight)
+
+    if pos_label is None:
+        pos_label = y_true.max()
+    y_true = np.array(y_true == pos_label, int)
+    y_true = _check_binary_probabilistic_predictions(y_true, y_prob)
+    step_size = 1 / float(nbins)
+    loss = 0.
+    count = 0.
+    for x in np.arange(0, 1, step_size):
+        in_range = (x <= y_prob) & (y_prob < x + step_size)
```

Comment:
You could get the bin for each y_prob with searchsorted, then use bincount (or ufunc.at) to get weighted sums

Reply:
Indeed, I was contemplating this option, or use of `np.digitize` and `np.bincount` as in the `calibration_curve` function.

Reply:
For a small number of bins, digitize is faster than searchsorted, but otherwise, which is chosen is immaterial